helix-core/src/main/java/org/apache/helix/task/TaskStateModelFactory.java

```diff
@@ -140,19 +139,18 @@ protected static RealmAwareZkClient createZkClient(HelixManager manager) {
     // and some connect the manager before registering the state model factory (in which case we
     // can use manager's connection). We need to think about the right order and determine if we
     // want to enforce it, which may cause backward incompatibility.
+    if (!(manager instanceof ZKHelixManager)) {
```

Comment:
Do we want to add a TODO here to indicate that we might need to extend the support in the future?

Reply:
I actually want to get an opinion on this: is it really better to reject non-ZKHelixManagers, comparing to returning a null here, and let the util function to use the default pool size? This means open source customers who extend HelixManager literally cannot create a TaskStateModelFactory.

Reply:
You can throw an [UnsupportedOperationException](https://docs.oracle.com/javase/6/docs/api/java/lang/UnsupportedOperationException.html) here and let the user know that only ZkHelixManager is currently supported. If we have different implementations of HelixManager down the road, you could add the support for them.

If there are custom implementations that extend HelixManager, they still can't create a ZkClient whether you have this change or not, so it's not making things any worse. Also, HelixManager interface already assumes a ZK-based implementation, so trying to accommodate some non-ZK HelixManager implementation out there that may or may not exist in this PR is a moot point. This is a good point to consider if we want to introduce a set of new interfaces that will replace existing data access interfaces though.

Reply:
I'll add a TODO here. The fact that non-ZK HelixManager simply cannot create this factory is pretty absurd, especially given that we have a default value that can be used. 

UnsupportedOperationException is a good suggestion, changing.